fm.c

```diff
@@ -669,219 +658,642 @@ void FM_HandleIO2(fm_t *chip)
+    chip->slot_sr[0][3][0] &= ~1;
+    chip->slot_sr[0][4][0] &= ~1;
+    chip->slot_sr[1][0][0] &= ~1;
+    chip->slot_sr[1][1][0] &= ~1;
+    chip->slot_sr[1][2][0] &= ~1;
+    chip->slot_sr[1][3][0] &= ~1;
+    chip->slot_sr[1][4][0] &= ~1;
+
+    // rr registers.
+    chip->slot_rr[0][0][0] &= ~1;
+    chip->slot_rr[0][1][0] &= ~1;
+    chip->slot_rr[0][2][0] &= ~1;
+    chip->slot_rr[0][3][0] &= ~1;
+    chip->slot_rr[1][0][0] &= ~1;
+    chip->slot_rr[1][1][0] &= ~1;
+    chip->slot_rr[1][2][0] &= ~1;
+    chip->slot_rr[1][3][0] &= ~1;
+
+    // sl registers.
+    chip->slot_sl[0][0][0] &= ~1;
+    chip->slot_sl[0][1][0] &= ~1;
+    chip->slot_sl[0][2][0] &= ~1;
+    chip->slot_sl[0][3][0] &= ~1;
+    chip->slot_sl[1][0][0] &= ~1;
+    chip->slot_sl[1][1][0] &= ~1;
+    chip->slot_sl[1][2][0] &= ~1;
+    chip->slot_sl[1][3][0] &= ~1;
+
+    // ssg-eg registers.
+    chip->slot_ssg_eg[0][0][0] &= ~1;
+    chip->slot_ssg_eg[0][1][0] &= ~1;
+    chip->slot_ssg_eg[0][2][0] &= ~1;
+    chip->slot_ssg_eg[0][3][0] &= ~1;
+    chip->slot_ssg_eg[1][0][0] &= ~1;
+    chip->slot_ssg_eg[1][1][0] &= ~1;
+    chip->slot_ssg_eg[1][2][0] &= ~1;
+    chip->slot_ssg_eg[1][3][0] &= ~1;
+
+    // fn low
+    chip->chan_fnum[0][0] &= ~1;
+    chip->chan_fnum[1][0] &= ~1;
+    chip->chan_fnum[2][0] &= ~1;
+    chip->chan_fnum[3][0] &= ~1;
+    chip->chan_fnum[4][0] &= ~1;
+    chip->chan_fnum[5][0] &= ~1;
+    chip->chan_fnum[6][0] &= ~1;
+    chip->chan_fnum[7][0] &= ~1;
+    chip->chan_fnum[8][0] &= ~1;
+    chip->chan_fnum[9][0] &= ~1;
+    chip->chan_fnum[10][0] &= ~1;
+
+    // fn low ch3
+    chip->chan_fnum_ch3[0][0] &= ~1;
+    chip->chan_fnum_ch3[1][0] &= ~1;
+    chip->chan_fnum_ch3[2][0] &= ~1;
+    chip->chan_fnum_ch3[3][0] &= ~1;
+    chip->chan_fnum_ch3[4][0] &= ~1;
+    chip->chan_fnum_ch3[5][0] &= ~1;
+    chip->chan_fnum_ch3[6][0] &= ~1;
+    chip->chan_fnum_ch3[7][0] &= ~1;
+    chip->chan_fnum_ch3[8][0] &= ~1;
+    chip->chan_fnum_ch3[9][0] &= ~1;
+    chip->chan_fnum_ch3[10][0] &= ~1;
+
+    // block fn high
+    chip->chan_block[0][0] &= ~1;
+    chip->chan_block[1][0] &= ~1;
+    chip->chan_block[2][0] &= ~1;
+
+    // block fn high ch3
+    chip->chan_block_ch3[0][0] &= ~1;
+    chip->chan_block_ch3[1][0] &= ~1;
+    chip->chan_block_ch3[2][0] &= ~1;
+
+    // connect
+    chip->chan_connect[0][0] &= ~1;
+    chip->chan_connect[1][0] &= ~1;
+    chip->chan_connect[2][0] &= ~1;
+
+    // fb
+    chip->chan_fb[0][0] &= ~1;
+    chip->chan_fb[1][0] &= ~1;
+    chip->chan_fb[2][0] &= ~1;
+
+    // pms
+    chip->chan_pms[0][0] &= ~1;
+    chip->chan_pms[1][0] &= ~1;
+    chip->chan_pms[2][0] &= ~1;
+
+    // ams
+    chip->chan_ams[0][0] &= ~1;
+    chip->chan_ams[1][0] &= ~1;
+
+    // pan
+    chip->chan_pan[0][0] &= ~1;
+    chip->chan_pan[1][0] &= ~1;
+
+    // a register
+    chip->chan_a4[0] = 0;
+    chip->chan_ac[0] = 0;
+}
+
+static inline void fm_rotate_registers(fm_t *chip, const int sel, const int from)

```

Comment:
What difference does it make when to rotate, here or as it was before. The readability has degraded.

Reply:
Previous method of rotation even when unrolled included 90-100 usages of CH_ROTATE and SLOT_ROTATE all of which is branching and cannot be predicted by the compiler. doing it as `fm_swap_registers` and `fm_rotate_registers` takes out the branching completely and uses a lookup table to determine which path to go for either swapping or rotating the values.

Reply:
I agree with you there. How about we redo the edits in small batches?